File: leetcode/1463_Cherry_Pickup_II.py
from typing import List
from unittest import TestCase

class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        dp = [[-1] * m for i in range(m)]
        dp[0][-1] = grid[0][0] + grid[0][-1]
        for row in grid[1:]:
            dp_cur = [[-1] * m for i in range(m)]
            for i in range(m):
                for j in range(m):
                    if dp[i][j] != -1:
                        for ni, nj in ((ni, nj) for ni in (i - 1, i, i + 1) for nj in (j - 1, j, j + 1) if ni >= 0 and ni < m and nj >= 0 and nj < m):
                            v = dp[i][j] + row[ni] + (row[nj] if ni != nj else 0)
                            if dp_cur[ni][nj] < v:
                                dp_cur[ni][nj] = v
            dp = dp_cur
        return max(map(max, dp))

class CherryPickupTest(TestCase):
    def test_1(self) -> None:
        self.assertEqual(Solution().cherryPickup([[3,1,1],[2,5,1],[1,5,5],[2,1,1]]), 24)

    def test_2(self) -> None:
        self.assertEqual(Solution().cherryPickup([[1,0,0,0,0,0,1],[2,0,0,0,0,3,0],[2,0,9,0,0,0,0],[0,3,0,5,4,0,0],[1,0,2,3,0,0,6]]), 28)

File: leetcode/1463_Cherry_Pickup_II.rs

struct Solution {}

impl Solution {
    pub fn cherry_pickup(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid[0].len();
        let mut dp = vec![vec![-1; m]; m];
        dp[0][m - 1] = grid[0][0] + grid[0][m - 1];
        for row in &grid[1..] {
            let mut dp_cur = vec![vec![-1; m]; m];
            for i in 0..m {
                for j in 0..m {
                    if dp[i][j] != -1 {
                        for &ni in [i as i32 - 1, i as i32, i as i32 + 1].iter() {
                            for &nj in [j as i32 - 1, j as i32, j as i32 + 1].iter() {
                                if ni >=0 && ni < m as i32 && nj >= 0 && nj < m as i32 {
                                    let v = dp[i][j] + row[ni as usize] + (if ni != nj { row[nj as usize] } else { 0 }); 
                                    if v > dp_cur[ni as usize][nj as usize] {
                                        dp_cur[ni as usize][nj as usize] = v;
                                    }

                                }
                            }
                        }
                    }
                }
            }
            dp = dp_cur;
        }
        *dp.iter().map(|a| a.iter().max().unwrap()).max().unwrap()
    }
}

mod cherry_pickup_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::cherry_pickup(vec![vec![3,1,1],vec![2,5,1],vec![1,5,5],vec![2,1,1]]), 24);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::cherry_pickup(vec![vec![1,0,0,0,0,0,1],vec![2,0,0,0,0,3,0],vec![2,0,9,0,0,0,0],vec![0,3,0,5,4,0,0],vec![1,0,2,3,0,0,6]]), 28);
    }

}